### ISSUE TITLE
Concept: Configurable behavior on event dispatching

### ISSUE THREAD

**Issue by antonkomarev** [cybercog/laravel-love#145]
How could we override what will happen on Love event will be dispatched?

This question was raised many times before. Each time I've returned to it a new ideas come into my mind but all of them have their pros and cons. Finally I've got all pieces together and ready to share with my vision.

Thanks to Pavel Volkov @volkv for helping me out with better solution.

First, we should encapsulate increment & decrement logic in jobs. Because we will be able to run those jobs from any place in any time without dispatching an event (which may be listened by many listeners at moment). Listeners will become very simple and synchronous but they will dispatch queued jobs.

Second, we should have more control over the queue. Because developers should be able to choose if they want to use stock package behavior or they want to modify it to their application requirements.

I've designed 4 different implementations of the second part. Each implementation will be described in separate comment.

*Spoiler: 4th solution will be implemented https://github.com/cybercog/laravel-love/issues/145#issuecomment-583730266*

**Comment by antonkomarev** [comment]
### 1. Love global jobs connection

Add `love.jobs_connection` config value which will affect on all jobs.

```php
return [
    // ...other config fields

    'jobs_connection' => null,
];
```

Right now its fine because we have only 2 listeners, but later it might be an issue when only some jobs will require to work synchronously.

**Pros:**

- Simple


**Cons:**

- Not extendable

**Comment by antonkomarev** [comment]
### 2. Queue options in config

Add `love.jobs` config array where we could define jobs queue options specific for each job.

```php
return [
    // ...other config fields
    'jobs' => [
        Cog\Laravel\Love\Reactant\Jobs\IncrementAggregatesJob => [
            'connection' => 'custom_connection_name',
            'queue' => 'custom_queue_name',
            'timeout' => 60,
            'delay' => 90,
        ],
        Cog\Laravel\Love\Reactant\Jobs\DecrementAggregatesJob => [
            'connection' => 'custom_connection_name',
            'queue' => 'custom_queue_name',
            'timeout' => 60,
            'delay' => 90,
        ],
    ],
];
```

**Pros:**

- Each job could be fine-tuned for the app requirements


**Cons:**

- One more place where breaking changes may pop up in future versions

**Comment by antonkomarev** [comment]
### 3. Love jobs config registry

Add `love.jobs` config array which will work as service locator.


```php
return  [
    // ...other config fields

    'jobs' => [
        Cog\Laravel\Love\Reactant\Jobs\IncrementAggregatesJob::class => App\Jobs\Love\ReactantIncrementAggregatesJob,
        Cog\Laravel\Love\Reactant\Jobs\DecrementAggregatesJob::class => App\Jobs\Love\ReactantDecrementAggregatesJob,
    ],
]
```

or

```php
return  [
    // ...other config fields

    'jobs' => [
        'reactant_increment_aggregates' => Cog\Laravel\Love\Reactant\Jobs\IncrementAggregatesJob::class,
        'reactant_decrement_aggregates' => Cog\Laravel\Love\Reactant\Jobs\DecrementAggregatesJob::class,
    ],
]
```

or 

```php
return  [
    // ...other config fields

    'jobs' => [
        'reactant' => [
            'increment_aggregates' => Cog\Laravel\Love\Reactant\Jobs\IncrementAggregatesJob::class,
            'decrement_aggregates' => Cog\Laravel\Love\Reactant\Jobs\DecrementAggregatesJob::class,
        ],
    ],
]
```

**Pros:**

- You can easily replace stock jobs with your own ones


**Cons:**

- For developer it's not obvious where this rebinding is being done
- Not obvious what arguments will be passed in job's __construct method
- One more place where breaking changes may pop up in future versions

**Comment by antonkomarev** [comment]
### 4. Love event service provider

Introduce `LoveEventServiceProvider` class which will be responsible for registering Event Listeners only.

```php
namespace Cog\Laravel\Love;

use Illuminate\Foundation\Support\Providers\EventServiceProvider;

final class LoveEventServiceProvider extends EventServiceProvider
{
    /**
     * The event listener mappings for the application.
     *
     * @var array
     */
    protected $listen = [
        ReactionHasBeenAdded::class => [
            IncrementAggregates::class,
        ],
        ReactionHasBeenRemoved::class => [
            DecrementAggregates::class,
        ],
    ];
}
```

**Pros:**

- Easy to opt-out


**Cons:**

- Breaking change

**Comment by antonkomarev** [comment]
This concept has been implemented in these PRs: #146 #147
